calico_node/startup.go

```diff
@@ -14,59 +14,657 @@
-// all of the function from startup.py will be moved here.
+// This file contains the main startup processing for the calico/node.  This
+// includes:
+// -  Detecting IP address and Network to use for BGP
+// -  Configuring the node resource with IP/AS information provided in the
+//    environment, or autodetected.
+// -  Creating default IP Pools for quick-start use
+// -  TODO:  Configuring IPIP tunnel with an IP address from an IP pool
+// TODO: Different auto-detection methods
+
 func main() {
-	// Build a Calico client.
-	log.Info("Creating Calico client")
+	var err error
+
+	// Determine the name for this node.
+	nodeName := determineNodeName()
+
+	// Create the Calico API client.
+	cfg, client := createClient()
+
+	// If this is a Kubernetes backed datastore then just make sure the
+	// datastore is initialized and then exit.  We don't need to explicitly
+	// initialize the datastore for non-Kubernetes because the node resource
+	// management will do that for us.
+	if cfg.Spec.DatastoreType == api.Kubernetes {
+		message("Calico is using a Kubernetes datastore")
+		err = client.EnsureInitialized()
+		if err != nil {
+			fatal("Error initializing Kubernetes as the datastore: %s", err)
+			terminate()
+		}
+		log.Info("Kubernetes is initialized as a Calico datastore")
+		writeStartupEnv(nodeName, "", "")
+		return
+	}
+
+	// Query the current Node resources.  We update our node resource with
+	// updated IP data and use the full list of nodes for validation.
+	node := getNode(client, nodeName)
+
+	// If running in policy only mode, apply the node resource and exit.
+	// This ensures we have an entry that can be queried, and also ensures
+	// the datastore is initialized.
+	if strings.ToLower(os.Getenv("CALICO_NETWORKING_BACKEND")) == "none" {
+		if _, err := client.Nodes().Apply(node); err != nil {
+			fatal("Unable to set node resource configuration: %s", err)
+			terminate()
+		}
+		message("Running calico/node in policy only mode")
+		writeStartupEnv(nodeName, "", "")
+		return
+	}
+
+	// Configure and verify the node IP addresses and subnets.
+	configureIPsAndSubnets(node)
+
+	// Configure the node AS number.
+	configureASNumber(node)
+
+	// Configure IP Pool configuration.
+	configureIPPools(client)
+
+	// Check for conflicting node configuration
+	checkConflictingNodes(client, node)
+
+	// Apply the updated node resource.
+	if _, err := client.Nodes().Apply(node); err != nil {
+		fatal("Unable to set node resource configuration: %s", err)
+		terminate()
+	}
+
+	// Set other Felix config that is not yet in the node resource.
+	if err := ensureDefaultConfig(client, node); err != nil {
+		fatal("Unable to set global default configuration: %s", err)
+		terminate()
+	}
+
+	// Write the startup.env file now that we are ready to start other
+	// components.
+	ipv6Str := ""
+	if node.Spec.BGP.IPv6Address != nil {
+		ipv6Str = node.Spec.BGP.IPv6Address.String()
+	}
+	writeStartupEnv(nodeName, node.Spec.BGP.IPv4Address.String(), ipv6Str)
```

Comment:
At the moment we always assume an IPv4 address is specified.  That's how it used to be and we hadn't yet changed that behavior.  I'll update the code though so that both or either can be nil -- I think I hacked this in when I was testing, so it does need a little work.